client/handler.go

```diff
@@ -122,6 +134,70 @@ func (h *Handler) Create(w http.ResponseWriter, r *http.Request, _ httprouter.Pa
 	h.r.Writer().WriteCreated(w, r, ClientsHandlerPath+"/"+c.GetID(), &c)
 }
 
+// swagger:route POST /dyn-clients public createDynOAuth2Client
+//
+// Create an OAuth 2.0 Client
+//
+// Create a new OAuth 2.0 client If you pass `client_secret` the secret will be used, otherwise a random secret will be generated. The secret will be returned in the response and you will not be able to retrieve it later on. Write the secret down and keep it somwhere safe.
+//
+// OAuth 2.0 clients are used to perform OAuth 2.0 and OpenID Connect flows. Usually, OAuth 2.0 clients are generated for applications which want to consume your OAuth 2.0 or OpenID Connect capabilities. To manage ORY Hydra, you will need an OAuth 2.0 Client as well. Make sure that this endpoint is well protected and only callable by first-party components.
+//
+//
+//     Consumes:
+//     - application/json
+//
+//     Produces:
+//     - application/json
+//
+//     Schemes: http, https
+//
+//     Responses:
+//       201: oAuth2Client
+//       400: genericError
+//       409: genericError
+//       500: genericError
+func (h *Handler) CreateDynamicRegistration(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	var c Client
+
```

Comment:
The OIDC spec requires authorization to perform dynamic client registration. It's not mandatory, but having this endpoint public could lead to DoS attacks on the DB by creating thousands or millions of clients.

Any ideas how we could/should protect this endpoint?

Also, in the OIDC spec they return a `registration_access_token` which can then be used to perform updates/reads on the client.

https://openid.net/specs/openid-connect-registration-1_0.html#RegistrationResponse

Reply:
@aeneasr  regarding how to prevent DoS attacks...  What I have seem in other situations is to rate limit the endpoint and when it is exceed the all next request return an error response during some time. Usually this is done in a higher level (Load balancer, api gateway or some routing engines) where the can implement this reate limit based on the referrer to do not completely block the endpoint for all users.

I think we can implement something similar. Something like add 2 configuration variables.. 1 For the rate limit and another for the timeframe to wait between the service is blocked until it accept requests again. Makes sense for you?

Reply:
Rate limiting is a difficult approach because it usually works on an IP basis and would require to be set up for every hydra installation as this endpoint is now exposed publicly. Alternatively, we could disable this endpoint per default and enable it per config which would mean that people have to add proper authorization themselves.

Reply:
@aeneasr Great... We have already a `SERVE_PUBLIC_ALLOW_DYNAMIC_REGISTRATION` configuration. Which is false be default.